docker/docker_client.go

```diff
@@ -533,6 +533,11 @@ func (c *dockerClient) getBearerToken(ctx context.Context, challenge challenge,
 	defer res.Body.Close()
 	switch res.StatusCode {
 	case http.StatusUnauthorized:
+		resBody, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Debugf("%s", string(resBody))
```

Comment:
Ideally, this should parse the body to extract the actual error message, if any, without all the cruft; from a quick look it seems that at least docker.io and registry.redhat.io (but not quay.io) use some kind of JSON format.

At the very least, the log should clearly say what this data is (“Server response when trying to obtain an access token: %s”), or something like that, and maybe deal with newlines.

Reply:
in that issue, `registry.gitlab.com` was used, from the log I saw different json format. I just update the log print format.

Reply:
So, looking at https://github.com/docker/distribution/blob/14b96e55d84c367ebab6caf9f0086de0876eec72/registry/client/auth/session.go#L441 , the body should be processed with `client.HandleErrorResponse`, like most registry API accesses. That accounts for both the docker.io and registry.redhat.io JSON responses I’ve seen.

Quay.io does not conform to that expectation, but that’s separately discussed in #703.

Reply:
Could we just wrap the error and continue to return the original errors.  Then Podman/Buildah could use errors.Cause and still get the detailed information.

Reply:
Not directly, because `ErrUnauthorizedForCredentials` is an error _value_; replacing it with an `errors.Wrap` would be a different value and an API break.

(Sure, we _can_ consider such an API break.)

Reply:
Let's break the API, for this, since this is useful information for the tools to report.

Reply:
Would have been nice delay the v4.0.0 for that :)